<a href="https://colab.research.google.com/github/Steve-YJ/Assignment_Standalone_DL/blob/master/%5BAgain%5D%5BResNet_VAE%5D_%5BJICS%5D_Exp01_Feature_Extraction_using_pre_trained_ResNet_VAE_Encoder__and_Classification_Ensemble_ipynb%EC%9D%98_%EC%82%AC%EB%B3%B8.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# README.MD
## Preparing JICS
* 🏅 Try as many as possible
* For Variant Malware Classification
    * Feature Extraction: Use pre-trained ResNet-VAE's Encoder Network
    * 📌 Malware Family Classification: Emsemble Model
    * 📌📌 Evaluation
        * ✅ Accuracy
        * ✅ precision, recall, F-1 Score etc...
        * 📌📌 Make Plot!!
* Experiment Details
    * 1. Load pre-trained ResNet-VAE's Encoder Network
        * Do-it: Feature Extraction using pre-trained ResNet-VAE's Encoder
    * 2. Data Preparation
        * Do-it: 
            * Feature Extraction
            * Data Preparation
            * Train_Test_Split
    * 3. Model Definition
        * Do-it: make Ensemble Classifier
    * 4. Evaluation
        * Do-it: Model Evaluation
            * Accuracy, precision, recall, F-1 Score, ROC ...
            
✅ Check Point<br>
"틈틈히 Refactoring 합시다...!!"
>  1. Load ResNet-VAE's 92th Model&Optim Parameters
>  2. Hyperparameter Optimization
>  3. Make Ensemble Classifier
>  4. 📌📌 Exp Report & Save Results

<code>from--- 20.09.28.mon am 00:30 ~ </code><br>
<code>Cont...--- 20.09.29.tue pm 11:30 ~ </code><br>
<code>Cont...--- 20.09.30.wed am 08:30 ~ </code>


<br>
✅ Check Point<br>

* Change Plan
   * Trying to Apply Hybrid Method with ResNet-VAE Feature Extractor -20.09.29.Tue-
    * ✅ 1. ResNet-VAE & Ensemble Classifier
    * 2. ResNet-VAE & MLP(Multi-Layer Perceptron)
    * 3. ResNet-VAE & CNN Classifier

> 🔖 Kaggle Kernel, git, tutorial 등을 참고해서 report 실험을 할 수 있다면 좋을 것 같습니다<br>
> ✅ 좀 더 Fancy한 모델을 통해 Accuracy, Precision, Recall, F-1 Score 등 평가 지수 높이기...!



---
 ✋ Hyperparameter Optimization(하이퍼 파라미터 최적화)는 조금 미뤄두자.<br></br>
우선은 계획대로 실험을 진행해보자...! -20.09.28.mon-
---

# Mount Drive



In [1]:
# Mount Drive
from google.colab import drive
drive.mount('/content/drive', force_remount=True)

Mounted at /content/drive


<code>Auto Reload</code>

In [2]:
%load_ext autoreload
%autoreload 2

In [3]:
%cd drive/My\ Drive/Post_InfoSec_Exps/ResNet-VAE/ResNetVAE-master.zip (Unzipped Files)/ResNetVAE-master
# ! ls

/content/drive/My Drive/Post_InfoSec_Exps/ResNet-VAE/ResNetVAE-master.zip (Unzipped Files)/ResNetVAE-master


In [4]:
import os 

def check_mkdir(dir_name):
    if not os.path.exists(dir_name):
        os.mkdir(dir_name)

In [5]:
check_mkdir('./ResNet-VAE_Exp2_3_1.ResNet-VAE_Encoder_Transfer_Learning')

In [6]:
! ls

'01.Tutorial-ResNet-VAE.ipynb의 사본'
 01.Tutorial-ResNet-VAE-Recon.ipynb
 02.Tutorial-ResNet-VAE-Tunning.ipynb
 03.Tutorial-ResNet-VAE-Tunning.ipynb
'03.Tutorial-ResNet-VAE-Tunning.ipynb의 사본'
'04.Post-01.Tutorial-ResNet-VAE.ipynb사본의 사본'
'05.Post-01.Tutorial-ResNet-VAE_Train_Again.ipynb의 사본'
 19train_val_plot.png
 39train_val_plot.png
 59train_val_plot.png
 Again_ResNet-VAE_Exp01
'[Again][ResNet-VAE] Exp2_3_1. Transfer Learning Using ResNet-VAE'\''Encoer.ipynb'
'[Again][ResNet-VAE] Exp2_3_1. Transfer Learning Using ResNet-VAE'\''Encoer.ipynb의 사본'
'[Again][ResNet-VAE] [JICS] Exp01. Feature_Extraction_using pre-trained ResNet-VAE Encoder _and_Classification_Ensemble .ipynb의 사본'
'[Again] ResNet-VAE_plus_Classifier .ipynb'
'[Again] ResNet-VAE_plus_Classifier .ipynb의 사본'
'[Again] Re-Start_Training_ResNet-VAE .ipynb'
 fig
 generated_Malimg.png
 modules.py
 plot_latent.ipynb
'[Post_Exp]04-2.ResNet-VAE_Train_Again.ipynb'
'[Post_Exp]04-2.ResNet-VAE_Train_Again.ipynbᄋ

In [7]:
save_path =  'ResNet-VAE_Exp2_3_1.ResNet-VAE_Encoder_Transfer_Learning'
save_path  # 실험 결과 저장 경로 설정

'ResNet-VAE_Exp2_3_1.ResNet-VAE_Encoder_Transfer_Learning'

# #01. Import Library

* pre-trained ResNet-VAE 모델을 정의하고 불러오기 위해 관련 라이브러리 import


In [8]:
from PIL import Image
%matplotlib inline

import matplotlib
import matplotlib.pyplot as plt
import matplotlib as mpl

import seaborn as sns

# save single numpy array
from tempfile import TemporaryFile
from sklearn.manifold import TSNE

import torch
import torch.utils.data  # torch.utils.data

from torch import nn, optim
from torch.nn import functional as F
from torch.autograd import Variable
from torch.utils.data import Dataset, DataLoader

import numpy as np
import pandas as pd

# argparse
import argparse
import time

import torchvision 
from torchvision import datasets, transforms
from torchvision.utils import save_image

# load modules
from torchvision import models
from modules import *

# #02. Data Preparation

In [9]:
transforms = transforms.Compose([
    transforms.Resize((224, 224)),
    transforms.Grayscale(num_output_channels=3), # make 3-channel Images
    transforms.ToTensor()])  # Composes several transforms together.

# make custom dataset
trainset = torchvision.datasets.ImageFolder(root='../../../../InformationSecurity_Summer/malimg',
                                            transform=transforms)

In [10]:
classes = trainset.classes
classes

['Adialer.C',
 'Agent.FYI',
 'Allaple.A',
 'Allaple.L',
 'Alueron.gen!J',
 'Autorun.K',
 'C2LOP.P',
 'C2LOP.gen!g',
 'Dialplatform.B',
 'Dontovo.A',
 'Fakerean',
 'Instantaccess',
 'Lolyda.AA1',
 'Lolyda.AA2',
 'Lolyda.AA3',
 'Lolyda.AT',
 'Malex.gen!J',
 'Obfuscator.AD',
 'Rbot!gen',
 'Skintrim.N',
 'Swizzor.gen!E',
 'Swizzor.gen!I',
 'VB.AT',
 'Wintrim.BX',
 'Yuner.A']

* Make Custom Dataset
    * Train_Val_Test split
<br>

❌ 향후 ResNet-VAE Encoder Feature Extraction & MLP 학습에서 사용가능한 데이터 셋이 완성되었다<br>
* Train: 7471
* Val: 933
* Test: 935
<br>

ResNet-VAE & MLP를 위해서는 뒤에서 다시 데이터셋을 생성해 줘야 한다 -20.09.30.wed- am 8:30

In [11]:
full_dataset = trainset

# maek train, val, test: 8:1:1
train_size = int(0.8 * len(full_dataset))
val_size = int(0.1 * len(full_dataset))
test_size = len(full_dataset) - train_size - val_size
print(train_size, val_size, test_size)

train_dataset, val_dataset, test_dataset = torch.utils.data.random_split(full_dataset, [train_size, val_size, test_size])

7471 933 935


pre-trained ResNet-VAE 모델의 모델 parameter를 가져오기 위해 해당 실험때와 같은 hyperparameter를 설정해준다

In [12]:
import argparse

# ====== Random Seed Initialization ====== #
seed = 123
np.random.seed(seed)  # Reseed a legacy MT19937 BitGenerator
torch.manual_seed(seed)

parser = argparse.ArgumentParser()
args = parser.parse_args("")

# ===== Model ====== #
args.CNN_fc_hidden1 = 1024
args.CNN_fc_hidden2 = 1024
args.CNN_embed_dim = 256  # latent dim extracted by 2D CNN
args.res_size = 224       # ResNet Image size

# ===== Regularization ===== #
args.dropout_p = 0.2           # dropout probability

# ===== training parameters ====== #
args.epochs = 20
args.learning_rate = 1e-3

args.batch_size = 50
args.log_interval = 10  # interval for displaying training info

print(args)

Namespace(CNN_embed_dim=256, CNN_fc_hidden1=1024, CNN_fc_hidden2=1024, batch_size=50, dropout_p=0.2, epochs=20, learning_rate=0.001, log_interval=10, res_size=224)


* z_loader's batch_size: 9339 -> 512

Feature Extraction을 위해 train, val, test외에 z_loader를 정의해준다

In [13]:
# make custom data_loader
train_loader = DataLoader(train_dataset,
                         batch_size=args.batch_size,
                         shuffle=True,
                         pin_memory=True)
val_loader = DataLoader(val_dataset,
                        batch_size=args.batch_size,
                        shuffle=True,
                        pin_memory=True)
 
test_loader = DataLoader(test_dataset,
                        batch_size=args.batch_size,
                        shuffle=True,
                        pin_memory=True)  # Instead, we recommend using automatic memory pinning (i.e., setting pin_memory=True)
                                          #  which enables fast data transfer to CUDA-enabled GPUs

# First, insert all test dataset
# z_loader: for latent vector extraction
z_loader = DataLoader(full_dataset,
                        batch_size=32,  # 128 -> 64 -> 32
                        shuffle=False,  # Shuffle -> False
                        pin_memory=True)

# #03. ResNet-VAE Architecture




## Make ResNet-VAE Architecture
* Referenced: https://github.com/hsinyilin19/ResNetVAE/blob/master/modules.py

실험 결과 저장 경로 설정 <code>save_model_path</code>

In [14]:
save_model_path = save_path
save_model_path

'ResNet-VAE_Exp2_3_1.ResNet-VAE_Encoder_Transfer_Learning'

In [15]:
def loss_function(recon_x, x, mu, logvar):
    # MSE = F.mse_loss(recon_x, x, reduction='sum')
    MSE = F.binary_cross_entropy(recon_x, x, reduction='sum')
    KLD = -0.5 * torch.sum(1 + logvar - mu.pow(2) - logvar.exp())
    return MSE + KLD

In [16]:
class ResNet_VAE(nn.Module):
    def __init__(self, fc_hidden1=1024, fc_hidden2=768, drop_p=0.3, CNN_embed_dim=256):
        super(ResNet_VAE, self).__init__()

        self.fc_hidden1, self.fc_hidden2, self.CNN_embed_dim = fc_hidden1, fc_hidden2, CNN_embed_dim

        # CNN architechtures
        self.ch1, self.ch2, self.ch3, self.ch4 = 16, 32, 64, 128
        self.k1, self.k2, self.k3, self.k4 = (5, 5), (3, 3), (3, 3), (3, 3)      # 2d kernal size
        self.s1, self.s2, self.s3, self.s4 = (2, 2), (2, 2), (2, 2), (2, 2)      # 2d strides
        self.pd1, self.pd2, self.pd3, self.pd4 = (0, 0), (0, 0), (0, 0), (0, 0)  # 2d padding

        # ====== Encoder ====== #
        # encoding components
        resnet = models.resnet152(pretrained=True)
        modules = list(resnet.children())[:-1]      # delete the last fc layer.
                                                    # pre-trained ResNet Architecture에서 마지막 FC레이어를 제외한 Model
        # ====== Notice ======= #
        # 이 코드를 잘 고치면 MLP Layer의 개수를 변경해 가면서 실험할 수 있겠다
        self.resnet = nn.Sequential(*modules)
        self.fc1 = nn.Linear(resnet.fc.in_features, self.fc_hidden1)
        self.bn1 = nn.BatchNorm1d(self.fc_hidden1, momentum=0.01)
        self.fc2 = nn.Linear(self.fc_hidden1, self.fc_hidden2)
        self.bn2 = nn.BatchNorm1d(self.fc_hidden2, momentum=0.01)

        # Latent vectors mu and sigma
        self.fc3_mu = nn.Linear(self.fc_hidden2, self.CNN_embed_dim)      # output = CNN embedding latent variables
        self.fc3_logvar = nn.Linear(self.fc_hidden2, self.CNN_embed_dim)  # output = CNN embedding latent variables

        # Sampling vector
        self.fc4 = nn.Linear(self.CNN_embed_dim, self.fc_hidden2)
        self.fc_bn4 = nn.BatchNorm1d(self.fc_hidden2)
        self.fc5 = nn.Linear(self.fc_hidden2, 64 * 4 * 4)
        self.fc_bn5 = nn.BatchNorm1d(64 * 4 * 4)
        self.relu = nn.ReLU(inplace=True)

        # ====== Decoder ====== #
        # Decoder
        self.convTrans6 = nn.Sequential(
            nn.ConvTranspose2d(in_channels=64, out_channels=32, kernel_size=self.k4, stride=self.s4,
                               padding=self.pd4),
            nn.BatchNorm2d(32, momentum=0.01),
            nn.ReLU(inplace=True),
        )
        self.convTrans7 = nn.Sequential(
            nn.ConvTranspose2d(in_channels=32, out_channels=8, kernel_size=self.k3, stride=self.s3,
                               padding=self.pd3),
            nn.BatchNorm2d(8, momentum=0.01),
            nn.ReLU(inplace=True),
        )

        self.convTrans8 = nn.Sequential(
            nn.ConvTranspose2d(in_channels=8, out_channels=3, kernel_size=self.k2, stride=self.s2,
                               padding=self.pd2),
            nn.BatchNorm2d(3, momentum=0.01),
            nn.Sigmoid()    # y = (y1, y2, y3) \in [0 ,1]^3
        )


    def encode(self, x):
        x = self.resnet(x)  # ResNet
        x = x.view(x.size(0), -1)  # flatten output of conv

        # FC layers
        x = self.bn1(self.fc1(x))
        x = self.relu(x)
        x = self.bn2(self.fc2(x))
        x = self.relu(x)
        # x = F.dropout(x, p=self.drop_p, training=self.training)
        mu, logvar = self.fc3_mu(x), self.fc3_logvar(x)
        return mu, logvar  # check: Encoder -> hid1-BN-hid2-BN ==> x

    def reparameterize(self, mu, logvar):
        if self.training:
            std = logvar.mul(0.5).exp_()
            eps = Variable(std.data.new(std.size()).normal_())
            return eps.mul(std).add_(mu)
        else:
            return mu

    def decode(self, z):
        x = self.relu(self.fc_bn4(self.fc4(z)))
        x = self.relu(self.fc_bn5(self.fc5(x))).view(-1, 64, 4, 4)
        x = self.convTrans6(x)
        x = self.convTrans7(x)
        x = self.convTrans8(x)
        x = F.interpolate(x, size=(224, 224), mode='bilinear')
        return x

    def forward(self, x):
        mu, logvar = self.encode(x)
        z = self.reparameterize(mu, logvar)
        x_reconst = self.decode(z)

        return x_reconst, z, mu, logvar

📌 Custom DataLoader의 경우, 자동화를 위해 Train, Val, Test 함수 내에 만들어준다

In [17]:
# Detect devices
use_cuda = torch.cuda.is_available()
print(use_cuda)
device = torch.device("cuda:0" if use_cuda else "cpu")
print(device)

True
cuda:0


95 Epoch이 최선이지만... 데이터를 지워버렸기에 ㅠ 차선인 Epoch 90의 값을 불러온다...!

아래의 셀을 실행하면 ./results_ResNet-VAE_Exp01에 저장되어있는 pre-trained model의 parameter값을 불러올 수 있다

In [18]:
pre_saved_model_path = './results_ResNet-VAE_Exp01'
os.path.join(pre_saved_model_path, 'model_epoch{}.pth'.format(90))

'./results_ResNet-VAE_Exp01/model_epoch90.pth'

* 90 Epoch의 모델, 옵티마이져 파라미터를 불러온다
* 이 때 모델의 Encoer Network를 통해 Feature Extraction만 진행할 경우 <code>model.eval()</code>을 꼭 해준다

In [19]:
### If you want to use pre-trained model ####
pre_saved_model_path = './results_ResNet-VAE_Exp01'
epoch=90

# Create model
resnet_vae = ResNet_VAE(fc_hidden1=args.CNN_fc_hidden1, fc_hidden2=args.CNN_fc_hidden2, drop_p=args.dropout_p, CNN_embed_dim=args.CNN_embed_dim).to(device)
### If you want to use pre-trained model ####
resnet_vae.load_state_dict(torch.load(os.path.join(pre_saved_model_path, 'model_epoch{}.pth'.format(epoch))))
resnet_vae.eval()  # 오늘 한 건했던 model.eval()

model_params = list(resnet_vae.parameters())
optimizer = torch.optim.Adam(model_params, lr=args.learning_rate)
### If you want to use pre-trained model's optimizer ####
optimizer.load_state_dict(torch.load(os.path.join(pre_saved_model_path, 'optimizer_epoch{}.pth'.format(epoch))))


print('Number of {} parameters'.format(sum(p.numel() for p in resnet_vae.parameters() if p.requires_grad)))
print("Using", torch.cuda.device_count(), "GPU!")

Downloading: "https://download.pytorch.org/models/resnet152-b121ed2d.pth" to /root/.cache/torch/hub/checkpoints/resnet152-b121ed2d.pth



Number of 63158425 parameters
Using 1 GPU!


In [20]:
resnet_vae

ResNet_VAE(
  (resnet): Sequential(
    (0): Conv2d(3, 64, kernel_size=(7, 7), stride=(2, 2), padding=(3, 3), bias=False)
    (1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (2): ReLU(inplace=True)
    (3): MaxPool2d(kernel_size=3, stride=2, padding=1, dilation=1, ceil_mode=False)
    (4): Sequential(
      (0): Bottleneck(
        (conv1): Conv2d(64, 64, kernel_size=(1, 1), stride=(1, 1), bias=False)
        (bn1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (conv2): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
        (bn2): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (conv3): Conv2d(64, 256, kernel_size=(1, 1), stride=(1, 1), bias=False)
        (bn3): BatchNorm2d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (relu): ReLU(inplace=True)
        (downsample): Sequential(
          (0): Conv2d(64

# #04. Train, Validation, Test


## Training Ensemble Classifier
After train ResNet-VAE, Use EncoderNet as Feature Extractor<br>
<br>
* Extract Latent Vector
* Improve Classifier

* pre-traned ResNet-VAE의 Encoder를 <code>Feature Extractor</code>로 활용하여 분류 문제를 풀어보도록 한다. <code>20.09.28.mon pm4:00</code>

* Referernce:
* 추가실험 List
    * Load VAE_State_Dict
    * Extract Latent vector & make Training set(for DL/ML)
    * => For all training data: (9339, 256)  / Latent dim
    * => Add Labels: (9339, 257)  # 256 feature 1 label
    * Make Pandas DataFrame
    * Make pd.DataFrame()
    * Save DataFrame() to csv
    * Make Classifier
* Classification
    * ML: Voting Classifier
    * DL: ...

# ResNet-VAE & Ensemble 

## Extract Latent Vector

* 적당한 알고리즘을 짜야겠구나...! -20.09.28.mon pm6:50-
* 📌 아래의 코드는 Feature Extraction을 수행할 때만 실행

✅ 데이터는 백업이 생명이다...!!!<br>
latent_vector 저장후 백업을 하도록 하자...!

In [27]:
for i, (data, _) in enumerate(z_loader):  # load_whole data(9339)
    data = data.to(device)
    # recon_batch, mu, logvar, z = resnet_vae(data)
    recon_batch, z, mu, logvar = resnet_vae(data)  # batch_size만큼 값 저장
    
    # save latent_vector
    latent_vector = z.detach().cpu().clone().numpy()  # change tensor type data to cpu().numpy()
                                                        # latent_vector_size: (num_of_data, 256)  # num_of_data, num_of_dim
    # save label
    label = _.detach().cpu().clone().numpy()
    label =label.reshape(1, -1)
    print(label.shape)

    # concat_data
    if i == 0:
        latent_z = np.concatenate((latent_vector, label.T), axis=1)  # latent_z: (9339, 101): (num_data, latent_z + label) 
        print(latent_z.shape)  # (32, 257): batch_num, features + label
    elif i > 0:
        latent_z = np.concatenate((latent_z,  np.concatenate((latent_vector, label.T), axis=1)), axis=0)
        print(latent_z.shape)


/usr/local/lib/python3.6/dist-packages/torch/nn/functional.py:3121: UserWarning: Default upsampling behavior when mode=bilinear is changed to align_corners=False since 0.4.0. Please specify align_corners=True if the old behavior is desired. See the documentation of nn.Upsample for details.
  "See the documentation of nn.Upsample for details.".format(mode))


(1, 32)
(32, 257)
(1, 32)
(64, 257)
(1, 32)
(96, 257)
(1, 32)
(128, 257)
(1, 32)
(160, 257)
(1, 32)
(192, 257)
(1, 32)
(224, 257)
(1, 32)
(256, 257)
(1, 32)
(288, 257)
(1, 32)
(320, 257)
(1, 32)
(352, 257)
(1, 32)
(384, 257)
(1, 32)
(416, 257)
(1, 32)
(448, 257)
(1, 32)
(480, 257)
(1, 32)
(512, 257)
(1, 32)
(544, 257)
(1, 32)
(576, 257)
(1, 32)
(608, 257)
(1, 32)
(640, 257)
(1, 32)
(672, 257)
(1, 32)
(704, 257)
(1, 32)
(736, 257)
(1, 32)
(768, 257)
(1, 32)
(800, 257)
(1, 32)
(832, 257)
(1, 32)
(864, 257)
(1, 32)
(896, 257)
(1, 32)
(928, 257)
(1, 32)
(960, 257)
(1, 32)
(992, 257)
(1, 32)
(1024, 257)
(1, 32)
(1056, 257)
(1, 32)
(1088, 257)
(1, 32)
(1120, 257)
(1, 32)
(1152, 257)
(1, 32)
(1184, 257)
(1, 32)
(1216, 257)
(1, 32)
(1248, 257)
(1, 32)
(1280, 257)
(1, 32)
(1312, 257)
(1, 32)
(1344, 257)
(1, 32)
(1376, 257)
(1, 32)
(1408, 257)
(1, 32)
(1440, 257)
(1, 32)
(1472, 257)
(1, 32)
(1504, 257)
(1, 32)
(1536, 257)
(1, 32)
(1568, 257)
(1, 32)
(1600, 257)
(1, 32)
(1632, 257)
(1, 32)
(1664,

In [28]:
latent_z.shape

(9339, 257)

save_model_path = 'ResNet-VAE_Exp2_3_1.ResNet-VAE_Encoder_Transfer_Learning'

백업해두자...!

In [29]:
with open(('./'+save_model_path+'/{}Epoch_z_vector_eval_2.npy'.format(epoch)), 'wb') as f:
        np.save(f, latent_z)  # but latent vector size is (16, 100).... just 16...
                                    # 1 Epoch 단위로 latent vector를 저장한다
                                    # 이 떄, latent vector의 size는 test_dataset의 크기가 되어야 한다
                                    # 결과값이 좋은 vector는 좋은 feature로 사용할 수 있다.
# plot latent vector Every 10 Epochs
print("Svae Latent vector!")

Svae Latent vector!


In [ ]:
latent_z.shape

## Data Preparation

📌 Feature Extraction이후 실험에서는 저장된 numpy값을 불러와서 사용하면 된다 -20.09.29.Tue-

In [23]:
! ls

'01.Tutorial-ResNet-VAE.ipynb의 사본'
 01.Tutorial-ResNet-VAE-Recon.ipynb
 02.Tutorial-ResNet-VAE-Tunning.ipynb
 03.Tutorial-ResNet-VAE-Tunning.ipynb
'03.Tutorial-ResNet-VAE-Tunning.ipynb의 사본'
'04.Post-01.Tutorial-ResNet-VAE.ipynb사본의 사본'
'05.Post-01.Tutorial-ResNet-VAE_Train_Again.ipynb의 사본'
 19train_val_plot.png
 39train_val_plot.png
 59train_val_plot.png
 Again_ResNet-VAE_Exp01
'[Again][ResNet-VAE] Exp2_3_1. Transfer Learning Using ResNet-VAE'\''Encoer.ipynb'
'[Again][ResNet-VAE] Exp2_3_1. Transfer Learning Using ResNet-VAE'\''Encoer.ipynb의 사본'
'[Again][ResNet-VAE] [JICS] Exp01. Feature_Extraction_using pre-trained ResNet-VAE Encoder _and_Classification_Ensemble .ipynb의 사본'
'[Again] ResNet-VAE_plus_Classifier .ipynb'
'[Again] ResNet-VAE_plus_Classifier .ipynb의 사본'
'[Again] Re-Start_Training_ResNet-VAE .ipynb'
 fig
 generated_Malimg.png
 modules.py
 plot_latent.ipynb
'[Post_Exp]04-2.ResNet-VAE_Train_Again.ipynb'
'[Post_Exp]04-2.ResNet-VAE_Train_Again.ipynbᄋ

In [30]:
print('./'+save_model_path+'/90Epoch_z_vector_eval_2.npy')
print(epoch)

./ResNet-VAE_Exp2_3_1.ResNet-VAE_Encoder_Transfer_Learning/90Epoch_z_vector_eval_2.npy
90


In [32]:
latent_vector = np.load('./'+save_model_path+'/{}Epoch_z_vector_eval_2.npy'.format(epoch))
print("Load Latent_Vector!!")
latent_vector.shape

Load Latent_Vector!!


(9339, 257)

In [ ]:
"""
data = latent_vector
print(data.shape)
"""

(9339, 257)


In [ ]:
"""
# numpy to pandas

import pandas as pd

data = pd.DataFrame(data=data)
data.head()
"""

0         1         2         3    ...       253       254       255  256
0 -0.079755 -0.147424 -0.189230  0.194458  ...  0.047721  0.031916 -0.250866  0.0
1 -0.075697 -0.141241 -0.202255  0.196442  ...  0.042003 -0.002187 -0.264128  0.0
2 -0.186946 -0.214237 -0.035928  0.453943  ...  0.095283  0.175777 -0.116401  0.0
3 -0.183893 -0.218137 -0.033342  0.455784  ...  0.100982  0.186238 -0.115219  0.0
4 -0.175223 -0.216755 -0.033997  0.452481  ...  0.107672  0.173136 -0.119256  0.0

[5 rows x 257 columns]

In [ ]:
"""
data.to_csv('./'+save_model_path+ '/ResNet-VAE_Latent_dataset_eval.csv', index=False)
df = pd.read_csv('./'+save_model_path+ '/ResNet-VAE_Latent_dataset_eval.csv')
df.head()
"""

## Ensemble Classifier


### 01. Vanila Voting Classifier

In [ ]:
# Test-Voting classifier
# https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html


from sklearn import datasets
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import GaussianNB
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import VotingClassifier

# iris = datasets.load_iris()
data = latent_vector  # re-change it to numpy
X, y = data[:, :256], data[:, 256]

clf1 = LogisticRegression(random_state=1)
clf2 = RandomForestClassifier(n_estimators=100, random_state=1)
clf3 = GaussianNB()

eclf = VotingClassifier(
         estimators=[('lr', clf1), ('rf', clf2), ('gnb', clf3)],
         voting='hard')

for clf, label in zip([clf1, clf2, clf3, eclf], ['Logistic Regression', 'Random Forest', 'naive Bayes', 'Ensemble']):
     scores = cross_val_score(clf, X, y, scoring='accuracy', cv=10)  # cross validation = 10
     print("Accuracy: %0.2f (+/- %0.2f) [%s]" % (scores.mean(), scores.std(), label))

/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression


Accuracy: 0.97 (+/- 0.01) [Logistic Regression]
Accuracy: 0.98 (+/- 0.00) [Random Forest]
Accuracy: 0.94 (+/- 0.01) [naive Bayes]


/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression


Accuracy: 0.98 (+/- 0.01) [Ensemble]


### [Tutorial] Voting Classifier
* reference: https://stackabuse.com/ensemble-voting-classification-in-python-with-scikit-learn/

In [ ]:
# importing libraries 
from sklearn.ensemble import VotingClassifier 
from sklearn.linear_model import LogisticRegression 
from sklearn.ensemble import RandomForestClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score, f1_score, log_loss, precision_score, recall_score, roc_auc_score
from sklearn.model_selection import train_test_split 
  
# loading iris dataset 
X, y = data[:, :256], data[:, 256]
print("Check X, y shape: ", X.shape, y.shape)
  
# train_test_split 
X_train, X_test, y_train, y_test = train_test_split(X,  
                                                    y,  
                                                    test_size = 0.1,  
                                                    random_state = 42) 
  
print('train_test_split')
print(X_train.shape, y_train.shape, X_test.shape, y_test.shape)  


# group / ensemble of models 
clf1 = LogisticRegression(random_state=1, multi_class ='multinomial')
clf2 = RandomForestClassifier(n_estimators=100, random_state=1)
clf3 = GaussianNB()

vot_hard = VotingClassifier(
         estimators=[('lr', clf1), ('rf', clf2), ('gnb', clf3)],
         voting='soft')

# Voting Classifier with hard voting 
vot_hard.fit(X_train, y_train) 
y_pred = vot_hard.predict(X_test)
  
# using accuracy_score metric to predict accuracy
"""
score = accuracy_score(y_test, y_pred) 
print('='*50)
print("Hard Voting Score % d" % score) 
print('='*50)
"""

# ===== print unique value y_test, y_pred ===== #
print('='*20, 'y_ unique', '='*20)
print(np.unique(y_pred), np.unique(y_test))


# Reference: https://medium.com/analytics-vidhya/voting-classifier-in-machine-learning-9534504eba39
for clf in (clf1, clf2, clf3, vot_hard):
    clf.fit(X_train, y_train)
    y_pred = clf.predict(X_test)
    print(clf.__class__.__name__, 'Accuracy: ', accuracy_score(y_test, y_pred))

    print(clf.__class__.__name__, 'Precision: ', precision_score(y_test,y_pred, 
                                               pos_label='positive',
                                               average='micro'))
    print(clf.__class__.__name__, 'Recall: ', recall_score(y_test, y_pred, 
                                          pos_label='positive',
                                          average='micro'))
    print(clf.__class__.__name__, 'ROC_AUC Score: ', roc_auc_score(y_test, clf.predict_proba(X_test),
                                                average='macro',
                                                multi_class='ovo'))
    
    
   

Check X, y shape:  (9339, 256) (9339,)
train_test_split
(8405, 256) (8405,) (934, 256) (934,)


/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


==================== y_ unique ====================
[ 0.  1.  2.  3.  4.  5.  6.  7.  8.  9. 10. 11. 12. 13. 14. 15. 16. 17.
 18. 19. 20. 21. 22. 23. 24.] [ 0.  1.  2.  3.  4.  5.  6.  7.  8.  9. 10. 11. 12. 13. 14. 15. 16. 17.
 18. 19. 20. 21. 22. 23. 24.]


/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)
/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1321: UserWarning: Note that pos_label (set to 'positive') is ignored when average != 'binary' (got 'micro'). You may use labels=[pos_label] to specify a single positive class.
  % (pos_label, average), UserWarning)


LogisticRegression Accuracy:  0.9753747323340471
LogisticRegression Precision:  0.9753747323340471
LogisticRegression Recall:  0.9753747323340471
LogisticRegression ROC_AUC Score:  0.9991517278016196
RandomForestClassifier Accuracy:  0.9817987152034261
RandomForestClassifier Precision:  0.9817987152034261
RandomForestClassifier Recall:  0.9817987152034261


/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1321: UserWarning: Note that pos_label (set to 'positive') is ignored when average != 'binary' (got 'micro'). You may use labels=[pos_label] to specify a single positive class.
  % (pos_label, average), UserWarning)


RandomForestClassifier ROC_AUC Score:  0.9987855113636362
GaussianNB Accuracy:  0.9379014989293362
GaussianNB Precision:  0.9379014989293362
GaussianNB Recall:  0.9379014989293362


/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1321: UserWarning: Note that pos_label (set to 'positive') is ignored when average != 'binary' (got 'micro'). You may use labels=[pos_label] to specify a single positive class.
  % (pos_label, average), UserWarning)


GaussianNB ROC_AUC Score:  0.9955280699069565


/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


VotingClassifier Accuracy:  0.9764453961456103
VotingClassifier Precision:  0.9764453961456103
VotingClassifier Recall:  0.9764453961456103


/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1321: UserWarning: Note that pos_label (set to 'positive') is ignored when average != 'binary' (got 'micro'). You may use labels=[pos_label] to specify a single positive class.
  % (pos_label, average), UserWarning)


VotingClassifier ROC_AUC Score:  0.9985725791754742


'\n# Voting Classifier with soft voting \neclf =  VotingClassifier(\n         estimators=[(\'lr\', clf1), (\'rf\', clf2), (\'gnb\', clf3)],\n         voting=\'soft\')\n\nfor clf, label in zip([clf1, clf2, clf3, eclf], [\'Logistic Regression\', \'Random Forest\', \'naive Bayes\', \'Ensemble\']):\n     scores = cross_val_score(clf, X_train, y_train, scoring=\'accuracy\', cv=10)  # cross validation = 10\n     print("Hard Voting Train Accuracy: %0.2f (+/- %0.2f) [%s]" % (scores.mean(), scores.std(), label))\nfor clf, label in zip([clf1, clf2, clf3, eclf], [\'Logistic Regression\', \'Random Forest\', \'naive Bayes\', \'Ensemble\']):\n     scores = cross_val_score(clf, X_test, y_test, scoring=\'accuracy\', cv=10)  # cross validation = 10\n     print("Hard Voting Accuracy: %0.2f (+/- %0.2f) [%s]" % (scores.mean(), scores.std(), label))\n\nvot_soft.fit(X_train, y_train) \ny_pred = vot_soft.predict(X_test) \n  \n# using accuracy_score \nscore = accuracy_score(y_test, y_pred) \nprint("Soft Votin

* Soft Voting시 Accuracy가 조금 떨어졌다

* Evaluation
    * VotingClassifier Accuracy:  0.9764453961456103
    * VotingClassifier Precision:  0.9764453961456103
    * VotingClassifier Recall:  0.9764453961456103
    * VotingClassifier ROC_AUC Score:  0. 9985725791754742


같은 hyperparameter에서 'hard voting'으로만 변경해보자<br>
단 hard voting은 ROC_AUC Score를 확인할 수 없다

In [ ]:
# importing libraries 
from sklearn.ensemble import VotingClassifier 
from sklearn.linear_model import LogisticRegression 
from sklearn.ensemble import RandomForestClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score, f1_score, log_loss, precision_score, recall_score, roc_auc_score
from sklearn.model_selection import train_test_split 
  
# loading iris dataset 
X, y = data[:, :256], data[:, 256]
print("Check X, y shape: ", X.shape, y.shape)
  
# train_test_split 
X_train, X_test, y_train, y_test = train_test_split(X,  
                                                    y,  
                                                    test_size = 0.1,  
                                                    random_state = 42) 
  
print('train_test_split')
print(X_train.shape, y_train.shape, X_test.shape, y_test.shape)  


# group / ensemble of models 
clf1 = LogisticRegression(random_state=1, multi_class ='multinomial')
clf2 = RandomForestClassifier(n_estimators=100, random_state=1)
clf3 = GaussianNB()

vot_hard = VotingClassifier(
         estimators=[('lr', clf1), ('rf', clf2), ('gnb', clf3)],
         voting='hard')  # soft to hard

# Voting Classifier with hard voting 
vot_hard.fit(X_train, y_train) 
y_pred = vot_hard.predict(X_test)
  
# using accuracy_score metric to predict accuracy


# ===== print unique value y_test, y_pred ===== #
print('='*20, 'y_ unique', '='*20)
print(np.unique(y_pred), np.unique(y_test))


# Reference: https://medium.com/analytics-vidhya/voting-classifier-in-machine-learning-9534504eba39
for clf in (clf1, clf2, clf3, vot_hard):
    clf.fit(X_train, y_train)
    y_pred = clf.predict(X_test)
    print(clf.__class__.__name__, 'Accuracy: ', accuracy_score(y_test, y_pred))

    print(clf.__class__.__name__, 'Precision: ', precision_score(y_test,y_pred, 
                                               pos_label='positive',
                                               average='micro'))
    print(clf.__class__.__name__, 'Recall: ', recall_score(y_test, y_pred, 
                                          pos_label='positive',
                                          average='micro'))
    
    # print(clf.__class__.__name__, 'ROC_AUC Score: ', roc_auc_score(y_test, clf.predict_proba(X_test),
    #                                             average='macro',
    #                                             multi_class='ovo'))
    
   

Check X, y shape:  (9339, 256) (9339,)
train_test_split
(8405, 256) (8405,) (934, 256) (934,)


/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


==================== y_ unique ====================
[ 0.  1.  2.  3.  4.  5.  6.  7.  8.  9. 10. 11. 12. 13. 14. 15. 16. 17.
 18. 19. 20. 21. 22. 23. 24.] [ 0.  1.  2.  3.  4.  5.  6.  7.  8.  9. 10. 11. 12. 13. 14. 15. 16. 17.
 18. 19. 20. 21. 22. 23. 24.]


/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)
/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1321: UserWarning: Note that pos_label (set to 'positive') is ignored when average != 'binary' (got 'micro'). You may use labels=[pos_label] to specify a single positive class.
  % (pos_label, average), UserWarning)


LogisticRegression Accuracy:  0.9753747323340471
LogisticRegression Precision:  0.9753747323340471
LogisticRegression Recall:  0.9753747323340471
RandomForestClassifier Accuracy:  0.9817987152034261
RandomForestClassifier Precision:  0.9817987152034261
RandomForestClassifier Recall:  0.9817987152034261
GaussianNB Accuracy:  0.9379014989293362
GaussianNB Precision:  0.9379014989293362
GaussianNB Recall:  0.9379014989293362


/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1321: UserWarning: Note that pos_label (set to 'positive') is ignored when average != 'binary' (got 'micro'). You may use labels=[pos_label] to specify a single positive class.
  % (pos_label, average), UserWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1321: UserWarning: Note that pos_label (set to 'positive') is ignored when average != 'binary' (got 'micro'). You may use labels=[pos_label] to specify a single positive class.
  % (pos_label, average), UserWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/

VotingClassifier Accuracy:  0.9817987152034261
VotingClassifier Precision:  0.9817987152034261
VotingClassifier Recall:  0.9817987152034261


/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1321: UserWarning: Note that pos_label (set to 'positive') is ignored when average != 'binary' (got 'micro'). You may use labels=[pos_label] to specify a single positive class.
  % (pos_label, average), UserWarning)


> VotingClassifier Accuracy:  0.9817987152034261<br>
> VotingClassifier Precision:  0.9817987152034261<br>
> VotingClassifier Recall:  0.9817987152034261

In [ ]:
# importing libraries 
from sklearn.ensemble import VotingClassifier 
from sklearn.linear_model import LogisticRegression 
from sklearn.svm import SVC 
from sklearn.tree import DecisionTreeClassifier 
from sklearn.datasets import load_iris 
from sklearn.metrics import accuracy_score 
from sklearn.model_selection import train_test_split 
  
# loading iris dataset 
X = data[:, 0:256]
y = data[:, 256]
print("Check X, y shape: ", X.shape, y.shape)
  
# train_test_split 
X_train, X_test, y_train, y_test = train_test_split(X,  
                                                    y,  
                                                    test_size = 0.20,  
                                                    random_state = 42) 
  
print('train_test_split')
print(X_train.shape, y_train.shape, X_test.shape, y_test.shape)  
# group / ensemble of models 
"""
estimator = [] 
estimator.append(('LR',  
                  LogisticRegression(solver ='lbfgs',  
                                     multi_class ='multinomial',  
                                     max_iter = 200))) 
estimator.append(('SVC', SVC(gamma ='auto', probability = True))) 
estimator.append(('DTC', DecisionTreeClassifier())) 
"""
  
clf1 = LogisticRegression(random_state=1)
clf2 = RandomForestClassifier(n_estimators=100, random_state=1)
clf3 = GaussianNB()

eclf = VotingClassifier(
         estimators=[('lr', clf1), ('rf', clf2), ('gnb', clf3)],
         voting='hard')

for clf, label in zip([clf1, clf2, clf3, eclf], ['Logistic Regression', 'Random Forest', 'naive Bayes', 'Ensemble']):
     scores = cross_val_score(clf, X_train, y_train, scoring='accuracy', cv=10)  # cross validation = 10
     print("Hard Voting Train Accuracy: %0.2f (+/- %0.2f) [%s]" % (scores.mean(), scores.std(), label))

for clf, label in zip([clf1, clf2, clf3, eclf], ['Logistic Regression', 'Random Forest', 'naive Bayes', 'Ensemble']):
     scores = cross_val_score(clf, X_test, y_test, scoring='accuracy', cv=10)  # cross validation = 10
     print("Hard Voting Test Accuracy: %0.2f (+/- %0.2f) [%s]" % (scores.mean(), scores.std(), label))
"""
# Voting Classifier with hard voting 
vot_hard.fit(X_train, y_train) 
y_pred = vot_hard.predict(X_test) 
  
# using accuracy_score metric to predict accuracy 
score = accuracy_score(y_test, y_pred) 
print("Hard Voting Score % d" % score) 
"""
# Voting Classifier with soft voting 
eclf =  VotingClassifier(
         estimators=[('lr', clf1), ('rf', clf2), ('gnb', clf3)],
         voting='soft')

for clf, label in zip([clf1, clf2, clf3, eclf], ['Logistic Regression', 'Random Forest', 'naive Bayes', 'Ensemble']):
     scores = cross_val_score(clf, X_train, y_train, scoring='accuracy', cv=10)  # cross validation = 10
     print("Hard Voting Train Accuracy: %0.2f (+/- %0.2f) [%s]" % (scores.mean(), scores.std(), label))
for clf, label in zip([clf1, clf2, clf3, eclf], ['Logistic Regression', 'Random Forest', 'naive Bayes', 'Ensemble']):
     scores = cross_val_score(clf, X_test, y_test, scoring='accuracy', cv=10)  # cross validation = 10
     print("Hard Voting Accuracy: %0.2f (+/- %0.2f) [%s]" % (scores.mean(), scores.std(), label))

"""
vot_soft.fit(X_train, y_train) 
y_pred = vot_soft.predict(X_test) 
  
# using accuracy_score 
score = accuracy_score(y_test, y_pred) 
print("Soft Voting Score % d" % score)
"""

Check X, y shape:  (9339, 256) (9339,)
train_test_split
(7471, 256) (7471,) (1868, 256) (1868,)


/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression


Hard Voting Train Accuracy: 0.97 (+/- 0.00) [Logistic Regression]
Hard Voting Train Accuracy: 0.98 (+/- 0.00) [Random Forest]
Hard Voting Train Accuracy: 0.94 (+/- 0.00) [naive Bayes]


/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression


Hard Voting Train Accuracy: 0.98 (+/- 0.00) [Ensemble]


/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression


Hard Voting Test Accuracy: 0.97 (+/- 0.01) [Logistic Regression]
Hard Voting Test Accuracy: 0.98 (+/- 0.01) [Random Forest]
Hard Voting Test Accuracy: 0.93 (+/- 0.02) [naive Bayes]


/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression


Hard Voting Test Accuracy: 0.98 (+/- 0.01) [Ensemble]


/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression


Hard Voting Train Accuracy: 0.97 (+/- 0.00) [Logistic Regression]
Hard Voting Train Accuracy: 0.98 (+/- 0.00) [Random Forest]
Hard Voting Train Accuracy: 0.94 (+/- 0.00) [naive Bayes]


/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression


Hard Voting Train Accuracy: 0.98 (+/- 0.00) [Ensemble]


/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression


Hard Voting Accuracy: 0.97 (+/- 0.01) [Logistic Regression]
Hard Voting Accuracy: 0.98 (+/- 0.01) [Random Forest]
Hard Voting Accuracy: 0.93 (+/- 0.02) [naive Bayes]


/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression


Hard Voting Accuracy: 0.97 (+/- 0.01) [Ensemble]


'\nvot_soft.fit(X_train, y_train) \ny_pred = vot_soft.predict(X_test) \n  \n# using accuracy_score \nscore = accuracy_score(y_test, y_pred) \nprint("Soft Voting Score % d" % score)\n'

### Voting Classifier3
* 5 RandomForest


In [ ]:
# Test-Voting classifier
# https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html


from sklearn import datasets
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import GaussianNB
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import VotingClassifier

# iris = datasets.load_iris()
data = latent_vector  # re-change it to numpy
X, y = data[:, :256], data[:, 256]
print('='*20,'Shape of X and y', '='*20)
print(X.shape, y.shape)
print('='*50)


clf1 = RandomForestClassifier(n_estimators=50, random_state=1)
clf2 = RandomForestClassifier(n_estimators=50, random_state=1)
clf3 = RandomForestClassifier(n_estimators=50, random_state=1)
clf4 = RandomForestClassifier(n_estimators=50, random_state=1)
clf5 = RandomForestClassifier(n_estimators=50, random_state=1)

eclf = VotingClassifier(
         estimators=[('lr', clf1), ('rf', clf2), ('gnb', clf3)],
         voting='hard')

for clf, label in zip([clf1, clf2, clf3, clf4, clf5,  eclf], ['Random Forest', 'Random Forest', 'Random Forest', 'Random Forest', 'Random Forest', 'Ensemble']):
     scores = cross_val_score(clf, X, y, scoring='accuracy', cv=10)  # cross validation = 10
     print("Accuracy: %0.2f (+/- %0.2f) [%s]" % (scores.mean(), scores.std(), label))
print("Done!!")

==================== Shape of X and y ====================
(9339, 256) (9339,)
Accuracy: 0.61 (+/- 0.02) [Random Forest]
Accuracy: 0.61 (+/- 0.02) [Random Forest]
Accuracy: 0.61 (+/- 0.02) [Random Forest]
Accuracy: 0.61 (+/- 0.02) [Random Forest]
Accuracy: 0.61 (+/- 0.02) [Random Forest]
Accuracy: 0.61 (+/- 0.02) [Ensemble]
Done!!


📌 최근에 Refactoring한 코드를 참고해서 본 연구의 실험 코드도 Refactoring 해보자..! <code>-20.09.28.Mon am 01:00</code>

### Voting Classifier4
* 👨‍🏫 1.11.6.4. Using the VotingClassifier with GridSearchCV
   * GridSearch를 통해 최적의 파라미터를 찾아보자
   * sklearn Documentation 참고

Voting Classifier can also be used together with GridSearchCV in order to tune the hyperparameters of the individual estimators

In [ ]:
from sklearn import datasets
from sklearn.model_selection import cross_val_score, GridSearchCV
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import GaussianNB
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import VotingClassifier

# iris = datasets.load_iris()
data = latent_vector  # re-change it to numpy
X, y = data[:, :256], data[:, 256]
print('='*20,'Shape of X and y', '='*20)
print(X.shape, y.shape)
print('='*50)


clf1 = LogisticRegression(random_state=1)
clf2 = RandomForestClassifier(random_state=7)
clf3 = GaussianNB()
eclf = VotingClassifier(
    estimators = [('lr', clf1), ('rf', clf2), ('gnb', clf3)],
    voting='soft'
)

params = {'lr__C': [1.0, 100.0], 'rf__n_estimators': [20, 200]}

grid = GridSearchCV(estimator=eclf, param_grid=params, cv=10)
grid = grid.fit(X, y)

==================== Shape of X and y ====================
(9339, 256) (9339,)


/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression


In [ ]:
for clf, label in zip([clf1, clf2, clf3, eclf], ['Logistic Regression', 'Random Forest', 'naive Bayes', 'Ensemble']):
    scores = cross_val_score(clf, X, y, scoring='accuracy', cv=10)
    print("Accuracy: %0.2f (+/- %0.2f) [%s]" % (scores.mean(), scores.std(), label))

/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression


Accuracy: 0.31 (+/- 0.01) [Logistic Regression]
Accuracy: 0.62 (+/- 0.03) [Random Forest]
Accuracy: 0.44 (+/- 0.01) [naive Bayes]


/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression


Accuracy: 0.50 (+/- 0.01) [Ensemble]


In [ ]:
from sklearn.neural_network import MLPClassifier

print(X.shape, y.shape)


clf = MLPClassifier(solver='sgd', 
                    alpha=1e-5,
                    # learning_rate=0.001,
                    max_iter=200,
                    early_stopping=True,
                    n_iter_no_change=100,
                     hidden_layer_sizes=(15,),
                     random_state=1)

clf.fit(X, y)


scores = cross_val_score(clf, X, y, scoring='accuracy', cv=10)
print(scores)

(9339, 256) (9339,)


/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


[0.48608137 0.49036403 0.52783726 0.51713062 0.51605996 0.51070664
 0.51498929 0.53104925 0.5267666  0.53376206]


/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


## ResNet-VAE Feature Extractor & MLP Classifier


* Process
    * pre-trained ResNet-VAE의 Encoder Netwrok로 부터 악성코드의 주요 특징을 추출한 데이터 셋을 구축한다
    * Custom Dataset, Custom DataLoader를 다시 구축한다
    * MLP 모델을 학습하며 각종 지표들을 검증해본다

<br>

✅ Re-Factoring Code를 참고하도록 한다...!(from. Stand-DL)

* From ResNet-VAE & Ensemble - Data Preparation
    * <code>latent_vector = np.load('./'+save_model_path+'/{}Epoch_z_vector_eval.npy'.format(epoch))
</code>

In [ ]:
latent_vector.shape

(9339, 257)

In [ ]:
X = latent_vector[:, :256]
y = latent_vector[:, 256]
print('print_shape: ', X.shape, y.shape)

print_shape:  (9339, 256) (9339,)


In [ ]:
from sklearn.model_selection import train_test_split


X_train, X_test, y_train, y_test = train_test_split(X,  
                                                    y,  
                                                    test_size = 0.20,  
                             å                       random_state = 42) 

아래의 train code는 ResNet-VAE train Code이다<br>
📌📌 => Encoder Network의 Train Code로 변경시켜보자...!<br>
* 참고하기
    * Reference: https://github.com/Steve-YJ/Assignment_Standalone_DL/blob/master/%5BRe_Fact%5D%20CIFAR-10_CNN_Report_Result.ipynb

* Encoder Network에서 <code>mu와 logvar</code>을 잘 활용해보자 -20.09.28.mon-
    * ResNet-152 -> FC1 -> FC2 -> mu or logvar

## Original Train Code - for ResNet-VAE Training

In [ ]:
def train(log_interval, model, device, train_loader, optimizer, epoch):
    # set model as training mode
    model.train()

    losses = []
    all_y, all_z, all_mu, all_logvar = [], [], [], []
    N_count = 0   # counting total trained sample in one epoch
    for batch_idx, (X, y) in enumerate(train_loader):
        # distribute data to device
        X, y = X.to(device), y.to(device).view(-1, )
        N_count += X.size(0)  # count batch_size sample

        optimizer.zero_grad()
        X_reconst, z, mu, logvar = model(X)  # VAE
        loss = loss_function(X_reconst, X, mu, logvar)
        losses.append(loss.item())

        loss.backward()
        optimizer.step()

        all_y.extend(y.data.cpu().numpy())
        all_z.extend(z.data.cpu().numpy())
        all_mu.extend(mu.data.cpu().numpy())
        all_logvar.extend(logvar.data.cpu().numpy())

        # show information
        if (batch_idx + 1) % log_interval == 0:  # if batch_size = 16 => 160
            print('Train Epoch: {} [{}/{} ({:.0f}%)]\tLoss: {:.6f}'.format(
                epoch + 1, N_count, len(train_loader.dataset), 100. * (batch_idx + 1) / len(train_loader), loss.item()))

    # calculate train_loss
    losses /= len(train_loader.dataset)
    all_y = np.stack(all_y, axis=0)
    all_z = np.stack(all_z, axis=0)
    all_mu = np.stack(all_mu, axis=0)
    all_logvar = np.stack(all_logvar, axis=0)

    # save Pytorch models of best record
    torch.save(model.state_dict(), os.path.join(save_model_path, 'model_epoch{}.pth'.format(epoch + 1)))  # save motion_encoder
    torch.save(optimizer.state_dict(), os.path.join(save_model_path, 'optimizer_epoch{}.pth'.format(epoch + 1)))      # save optimizer
    print("Epoch {} model saved!".format(epoch + 1))

    return X.data.cpu().numpy(), all_y, all_z, all_mu, all_logvar, losses



## Train Function for EncoderNet

* <code>Refactoring</code> Train Function for EncoderNet

In [ ]:
def train(net, trainloader, optimizer, criterion, args):
    
    """
    이미 위에서 정의를 해줌
    trainloader = torch.utils.data.DataLoader(partition['train'], 
                                              batch_size=args.train_batch_size, 
                                              shuffle=True, num_workers=2)
    """

    net.train()  # train mode
    optimizer.zero_grad()  # optimizer 초기화

    correct = 0
    total = 0
    train_loss = 0.0
    for i, data in enumerate(trainloader, 0):
        # get the inputs
        inputs, labels = data
        # print('input dimension: ', inputs.shape)
        # raise RuntimeError
        # inputs = inputs.view(-1, 3072)
        inputs = inputs.cuda()
        labels = labels.cuda()
        outputs = net.encoder(inputs)
        

        loss = criterion(outputs, labels)
        loss.backward()
        optimizer.step()

        train_loss += loss.item()
        _, predicted = torch.max(outputs.data, 1)
        total += labels.size(0)
        correct += (predicted == labels).sum().item()

    train_loss = train_loss / len(trainloader)
    train_acc = 100 * correct / total
    return net, train_loss, train_acc

In [ ]:
def val(net, valloader, optimizer, criterion, args):
    
    """
    이미 위에서 정의를 해줌
    valloader = torch.utils.data.DataLoader(partition['train'], 
                                              batch_size=args.test_batch_size, 
                                              shuffle=True, num_workers=2)
    """

    net.eval()

    correct = 0
    total = 0
    val_loss = 0.0

    with torch.no_grad():
        for i, data in enumerate(valloader, 0):
            # get the inputs
            inputs, labels = data
            # print('input dimension: ', inputs.shape)
            # raise RuntimeError
            # inputs = inputs.view(-1, 3072)
            inputs = inputs.cuda()
            labels = labels.cuda()
            outputs = net(inputs)

            loss = criterion(outputs, labels)
            
            val_loss += loss.item()
            _, predicted = torch.max(outputs.data, 1)
            total += labels.size(0)
            correct += (predicted == labels).sum().item()

        val_loss = val_loss / len(valloader)
        val_acc = 100 * correct / total
    return val_loss, val_acc

In [ ]:
def test(net, testloader, args):
    
    """
    이미 위에서 정의를 해줌
    testloader = torch.utils.data.DataLoader(partition['train'], 
                                              batch_size=args.test_batch_size, 
                                              shuffle=True, num_workers=2)
    """

    net.eval()

    correct = 0
    total = 0

    with torch.no_grad():
        for i, data in enumerate(testloader, 0):
            # get the inputs
            inputs, labels = data
            # print('input dimension: ', inputs.shape)
            # raise RuntimeError
            # inputs = inputs.view(-1, 3072)
            inputs = inputs.cuda()
            labels = labels.cuda()
            outputs = net(inputs)
            
            _, predicted = torch.max(outputs.data, 1)
            total += labels.size(0)
            correct += (predicted == labels).sum().item()

        test_acc = 100 * correct / total
    return test_acc

In [ ]:
def experiment(trainloader, valloader, testloader, args):

    """
    net = CNN(model_code = args.model_code,
              in_channels = args.in_channels,
              out_dim = args.out_dim,
              act = args.act,
              use_bn = args.use_bn)
    net.cuda()
    """
    # Detect devices
    use_cuda = torch.cuda.is_available()
    print(use_cuda)
    device = torch.device("cuda:0" if use_cuda else "cpu")
    print(device)

    """ 
    # ====== If you want to use pre-trained model ====== #
    """
    pre_saved_model_path = './results_ResNet-VAE_Exp01'
    epoch=90  # pre-trained ResNet-VAE parameter

    # Create model
    resnet_vae = ResNet_VAE(fc_hidden1=args.CNN_fc_hidden1, fc_hidden2=args.CNN_fc_hidden2, drop_p=args.dropout_p, CNN_embed_dim=args.CNN_embed_dim).to(device)
    ### If you want to use pre-trained model ####
    resnet_vae.load_state_dict(torch.load(os.path.join(pre_saved_model_path, 'model_epoch{}.pth'.format(epoch))))
    
    """
    # resnet_vae에서 encodernet을 어떻게 사용할 수 있을까? -20.09.28.mon-
    """
    


    model_params = list(resnet_vae.parameters())
    optimizer = torch.optim.Adam(model_params, lr=args.learning_rate)
    ### If you want to use pre-trained model's optimizer ####
    optimizer.load_state_dict(torch.load(os.path.join(pre_saved_model_path, 'optimizer_epoch{}.pth'.format(epoch))))


    print('Number of {} parameters'.format(sum(p.numel() for p in resnet_vae.parameters() if p.requires_grad)))
    print("Using", torch.cuda.device_count(), "GPU!")


    criterion = nn.CrossEntropyLoss()
    if args.optim == 'SGD':
        optimizer = optim.SGD(net.parameters(), lr=args.lr, weight_decay=args.l2)
    elif args.optim == 'RMSprop':
        optimizer = optim.RMSprop(net.parameters(), lr=args.lr, weight_decay=args.l2)
    elif args.optim == 'Adam':
        optimizer = optim.Adam(net.parameters(), lr=args.lr, weight_decay=args.l2)
    else:
        raise ValueError('In-valid optimizer choice')
    
    # ===== List for epoch-wise data ====== #
    train_losses = []
    val_losses = []
    train_accs = []
    val_accs = []
    # ================= Training ==================== #
        
    for epoch in range(args.epoch):  # loop over the dataset multiple times
        ts = time.time()
        net, train_loss, train_acc = train(net, trainloader, optimizer, criterion, args)
        val_loss, val_acc = validate(net, valloader, criterion, args)
        te = time.time()
        
        # ====== Add Epoch Data ====== #
        train_losses.append(train_loss)
        val_losses.append(val_loss)
        train_accs.append(train_acc)
        val_accs.append(val_acc)
        # ============================ #
        
        print('Epoch {}, Acc(train/val): {:2.2f}/{:2.2f}, Loss(train/val) {:2.2f}/{:2.2f}. Took {:2.2f} sec'.format(epoch, train_acc, val_acc, train_loss, val_loss, te-ts))
        
    test_acc = test(net, testloader, args)    
    
    # ======= Add Result to Dictionary ======= #
    # 전체 결과값을 저장하는 Results Dic 
    # Q. 왜 때문인진 모르겠지만 train_accs와 train_acc가 분리되어있네?!
        # train_acc와 val_acc은 없어도 되는거 아냐?
    result = {}
    result['train_losses'] = train_losses
    result['val_losses'] = val_losses
    result['train_accs'] = train_accs
    result['val_accs'] = val_accs
    # result['train_acc'] = train_acc
    # result['val_acc'] = val_acc
    result['test_acc'] = test_acc
    return vars(args), result  # vars(args)를 해주면 args를 dic로 저장해준다
    # ===================================== #

# #05. Manage Experiment

In [ ]:
import hashlib
import json
from os import listdir
from os.path import isfile, join
import pandas as pd

def save_exp_result(setting, result):  # setting: args
                                       # result : train_losses, val_losses, train_accs, val_accs,  
    exp_name = setting['exp_name']
    del setting['epoch']               # setting에서 'epoch'과 'test_batch_size'를 제거해준다고...
    del setting['test_batch_size']     # 왜 떼는거지?

    hash_key = hashlib.sha1(str(setting).encode()).hexdigest()[:6]  # setting에 따라 다른 파일명을 갖도록 만들어준다
    filename = './'+save_model_path+'/{}-{}.json'.format(exp_name, hash_key)
    # print(filename)
    result.update(setting)  # 결과 dictionary에 setting값을 더해준다 => Dic
    with open(filename, 'w') as f:
        json.dump(result, f)  # JSON 포멧으로 dictionary 값 저장

    
def load_exp_result(exp_name):
    dir_path = './results'
    filenames = [f for f in listdir(dir_path) if isfile(join(dir_path, f)) if '.json' in f]  # ./resuts에 저장되어있는 file들을 리스트 형태로 저장
    list_result = []
    for filename in filenames:
        if exp_name in filename:
            # print(exp_name)
            with open(join(dir_path, filename), 'r') as infile:
                results = json.load(infile)
                list_result.append(results)  # Dic을 list형태로 저장
    df = pd.DataFrame(list_result) # .drop(columns=[])
    return df


# #06. Experiment


In [ ]:
# ====== Random Seed Initialization ====== #
seed = 123  # 랜덤 시드값 초기화
np.random.seed(seed)
torch.manual_seed(seed)

parser = argparse.ArgumentParser()
args = parser.parse_args("")
args.exp_name = "exp1_lr_model_code"

# ====== Model ====== #
# args.model_code ='VGG11'
# args.in_channels = 3
args.out_dim = 10
# args.hid_dim = 100
args.act = 'relu'
args.CNN_fc_hidden1 = 1024
args.CNN_fc_hidden2 = 1024
args.CNN_embed_dim = 256  # latent dim extracted by 2D CNN
args.res_size = 224       # ResNet Image size
args.dropout_p = 0.2           # dropout probability



# ====== Regularization ======= #
# args.l2 = 0.00001
# args.use_bn = True

# ====== Optimizer & Training ====== #
args.epochs = 20
args.learning_rate = 1e-3

args.batch_size = 50
args.log_interval = 10  # interval for displaying training info

print(args)



# ===== Experiment Variable ====== #
# Reference: https://github.com/Steve-YJ/Assignment_Standalone_DL/blob/master/%5BRe_Fact%5D%20CIFAR-10_CNN_Report_Result.ipynb

name_var1 = 'lr'
name_var2 = 'hiden_layer_num'
list_var1 = [1e-3, 1e-4, 1e-5]
list_var2 = [1024, 512, 256, 128, 64, 32]


for var1 in list_var1:
    for var2 in list_var2:
        # setattr ??: name_var1('n_layer')를 1, 2, 3으로 바꿔준다
        # setattr = args.name_var1 = var1
        setattr(args, name_var1, var1)
        setattr(args, name_var2, var2)
        print(args)
                
        setting, result = experiment(train_loader, val_loader, test_loader, args)
        save_exp_result(setting, result)

Namespace(CNN_embed_dim=256, CNN_fc_hidden1=1024, CNN_fc_hidden2=1024, act='relu', batch_size=50, dropout_p=0.2, epochs=20, exp_name='exp1_lr_model_code', learning_rate=0.001, log_interval=10, out_dim=10, res_size=224)
Namespace(CNN_embed_dim=256, CNN_fc_hidden1=1024, CNN_fc_hidden2=1024, act='relu', batch_size=50, dropout_p=0.2, epochs=20, exp_name='exp1_lr_model_code', hiden_layer_num=1024, learning_rate=0.001, log_interval=10, lr=0.001, out_dim=10, res_size=224)
True
cuda:0


TypeError: ignored

# #07. Report Result